File: Daily/PY/Leetcode31-下一个排列.py
from typing import List
class Solution:
    def nextPermutation(self, nums: List[int]) -> None:
        #如何不越界的找到山峰
        i = len(nums) - 2
        while(i >= 0 and nums[i] >= nums[i+1]):
            i -= 1
        #可能i没有合适的值
        if(i >= 0):
            k = len(nums) - 1
            while(k >= i+1 and nums[k] <= nums[i]):
                k-=1
            nums[i],nums[k] = nums[k],nums[i]
        left = i+1
        right = len(nums) - 1
        #一个swap
        while left <= right:
            nums[left] ,nums[right] = nums[right],nums[left]
            left += 1
            right -= 1 

File: Leetcode/sbgo/Leetcode31-下一个排列.go
package sbgo

func nextPermutation(nums []int) {
	n := len(nums)
	i := n - 2
	for i >= 0 && nums[i] >= nums[i+1] {
		i--
	}
	if i >= 0 {
		j := n - 1
		for j >= i+1 && nums[j] <= nums[i] {
			j--
		}
		nums[i], nums[j] = nums[j], nums[i]
	}
	left, right := i+1, n-1
	for left < right {
		nums[left], nums[right] = nums[right], nums[left]
		left++
		right--
	}
}
